File: languages/compile-po-to-mo.php
#!/usr/bin/env php
<?php
/**
 * PO to MO Compiler
 * Converts PO translation files to MO binary format for WordPress
 */

if (php_sapi_name() !== 'cli') {
    die('This script must be run from the command line.');
}

class POtoMOCompiler {
    private $po_file;
    private $mo_file;
    
    public function __construct($po_file, $mo_file = null) {
        $this->po_file = $po_file;
        $this->mo_file = $mo_file ?: str_replace('.po', '.mo', $po_file);
    }
    
    public function compile() {
        if (!file_exists($this->po_file)) {
            throw new Exception("PO file not found: {$this->po_file}");
        }
        
        $translations = $this->parse_po_file();
        $this->write_mo_file($translations);
        
        return $this->mo_file;
    }
    
    private function parse_po_file() {
        $content = file_get_contents($this->po_file);
        $translations = [];
        
        // Parse PO file
        preg_match_all('/msgid\s+"(.*)"\s+msgstr\s+"(.*)"/Us', $content, $matches, PREG_SET_ORDER);
        
        foreach ($matches as $match) {
            $msgid = $this->unescape_string($match[1]);
            $msgstr = $this->unescape_string($match[2]);
            
            // Skip empty msgid (header) and empty translations
            if ($msgid !== '' && $msgstr !== '') {
                $translations[$msgid] = $msgstr;
            }
        }
        
        return $translations;
    }
    
    private function unescape_string($string) {
        $string = str_replace(['\\"', '\\n', '\\t', '\\r'], ['"', "\n", "\t", "\r"], $string);
        return $string;
    }
    
    private function write_mo_file($translations) {
        $count = count($translations);
        
        // MO file format (GNU gettext)
        $magic = 0x950412de; // Little endian
        $revision = 0;
        
        $ids = '';
        $strings = '';
        $id_offsets = [];
        $str_offsets = [];
        
        foreach ($translations as $id => $str) {
            $id_offsets[] = [strlen($ids), strlen($id)];
            $ids .= $id . "\0";
            
            $str_offsets[] = [strlen($strings), strlen($str)];
            $strings .= $str . "\0";
        }
        
        // Calculate offsets
        $key_start = 28;
        $value_start = $key_start + ($count * 8);
        $id_start = $value_start + ($count * 8);
        $str_start = $id_start + strlen($ids);
        
        // Build binary data
        $mo = pack('Iiiiiii',
            $magic,          // Magic number
            $revision,       // File format revision
            $count,          // Number of strings
            $key_start,      // Offset of table with original strings
            $value_start,    // Offset of table with translated strings
            0,               // Size of hashing table
            0                // Offset of hashing table
        );
        
        // Write original string offsets
        foreach ($id_offsets as $offset) {
            $mo .= pack('ii', $offset[1], $id_start + $offset[0]);
        }
        
        // Write translated string offsets
        foreach ($str_offsets as $offset) {
            $mo .= pack('ii', $offset[1], $str_start + $offset[0]);
        }
        
        // Write original strings
        $mo .= $ids;
        
        // Write translated strings
        $mo .= $strings;
        
        // Write to file
        if (file_put_contents($this->mo_file, $mo) === false) {
            throw new Exception("Failed to write MO file: {$this->mo_file}");
        }
        
        return true;
    }
}

// Main execution
try {
    $languages_dir = __DIR__;
    $files = [
        'azure-ai-chatbot-ko_KR.po',
        'azure-ai-chatbot-en_US.po'
    ];
    
    echo "========================================\n";
    echo "PO to MO Compiler for Azure AI Chatbot\n";
    echo "========================================\n\n";
    
    $compiled = 0;
    foreach ($files as $file) {
        $po_path = $languages_dir . '/' . $file;
        $mo_path = str_replace('.po', '.mo', $po_path);
        
        if (!file_exists($po_path)) {
            echo "⏭️  Skipping: $file (not found)\n";
            continue;
        }
        
        echo "📝 Compiling: $file\n";
        
        try {
            $compiler = new POtoMOCompiler($po_path, $mo_path);
            $result = $compiler->compile();
            
            $size = filesize($mo_path);
            echo "✅ Created: " . basename($result) . " (" . number_format($size) . " bytes)\n";
            $compiled++;
        } catch (Exception $e) {
            echo "❌ Error: " . $e->getMessage() . "\n";
        }
        
        echo "\n";
    }
    
    echo "========================================\n";
    echo "Compilation completed!\n";
    echo "Total compiled: $compiled file(s)\n";
    echo "========================================\n";
    
    exit(0);
    
} catch (Exception $e) {
    echo "Fatal error: " . $e->getMessage() . "\n";
    exit(1);
}

File: languages/compile-po-to-mo.py
#!/usr/bin/env python3
"""
PO to MO Compiler for WordPress
Converts PO translation files to MO binary format
"""

import struct
import os
import sys

def parse_po_file(po_path):
    """Parse PO file and extract translations"""
    translations = {}
    
    with open(po_path, 'r', encoding='utf-8') as f:
        content = f.read()
    
    # Simple parser for msgid/msgstr pairs
    lines = content.split('\n')
    i = 0
    
    while i < len(lines):
        line = lines[i].strip()
        
        # Skip comments and empty lines
        if not line or line.startswith('#'):
            i += 1
            continue
        
        # Look for msgid
        if line.startswith('msgid'):
            # Extract msgid
            msgid = line[6:].strip(' "')
            i += 1
            
            # Continue reading multi-line msgid
            while i < len(lines) and lines[i].strip().startswith('"'):
                msgid += lines[i].strip(' "')
                i += 1
            
            # Look for msgstr
            if i < len(lines) and lines[i].strip().startswith('msgstr'):
                msgstr = lines[i].strip()[7:].strip(' "')
                i += 1
                
                # Continue reading multi-line msgstr
                while i < len(lines) and lines[i].strip().startswith('"'):
                    msgstr += lines[i].strip(' "')
                    i += 1
                
                # Unescape strings
                msgid = msgid.replace('\\n', '\n').replace('\\t', '\t').replace('\\"', '"')
                msgstr = msgstr.replace('\\n', '\n').replace('\\t', '\t').replace('\\"', '"')
                
                # Add to translations (skip empty msgid and msgstr)
                if msgid and msgstr:
                    translations[msgid] = msgstr
        else:
            i += 1
    
    return translations

def write_mo_file(mo_path, translations):
    """Write translations to MO file"""
    
    # MO file format constants
    MAGIC = 0x950412de  # Little endian magic number
    VERSION = 0
    
    # Prepare strings
    keys = sorted(translations.keys())
    ids = b''.join(k.encode('utf-8') + b'\x00' for k in keys)
    strs = b''.join(translations[k].encode('utf-8') + b'\x00' for k in keys)
    
    # Calculate offsets
    key_start = 7 * 4 + 16 * len(keys)  # Header + offset table
    val_start = key_start + len(ids)
    
    # Build offset tables
    key_offsets = []
    str_offsets = []
    
    ids_offset = 0
    strs_offset = 0
    
    for key in keys:
        key_len = len(key.encode('utf-8'))
        str_len = len(translations[key].encode('utf-8'))
        
        key_offsets.append((key_len, key_start + ids_offset))
        str_offsets.append((str_len, val_start + strs_offset))
        
        ids_offset += key_len + 1
        strs_offset += str_len + 1
    
    # Build MO file
    mo = struct.pack(
        '<7I',  # Little endian, 7 unsigned integers
        MAGIC,              # Magic number
        VERSION,            # Version
        len(keys),          # Number of strings
        28,                 # Offset of original string table
        28 + len(keys) * 8, # Offset of translated string table
        0,                  # Hash table size
        0                   # Hash table offset
    )
    
    # Add key offset table
    for length, offset in key_offsets:
        mo += struct.pack('<2I', length, offset)
    
    # Add string offset table
    for length, offset in str_offsets:
        mo += struct.pack('<2I', length, offset)
    
    # Add strings
    mo += ids + strs
    
    # Write file
    with open(mo_path, 'wb') as f:
        f.write(mo)

def main():
    print('========================================')
    print('PO to MO Compiler for Azure AI Chatbot')
    print('========================================\n')
    
    # Get script directory
    script_dir = os.path.dirname(os.path.abspath(__file__))
    
    # Files to compile
    files = [
        'azure-ai-chatbot-ko_KR.po',
        'azure-ai-chatbot-en_US.po'
    ]
    
    compiled = 0
    
    for po_file in files:
        po_path = os.path.join(script_dir, po_file)
        mo_path = po_path.replace('.po', '.mo')
        
        if not os.path.exists(po_path):
            print(f'⏭️  Skipping: {po_file} (not found)')
            continue
        
        print(f'📝 Compiling: {po_file}')
        
        try:
            translations = parse_po_file(po_path)
            write_mo_file(mo_path, translations)
            
            size = os.path.getsize(mo_path)
            mo_name = os.path.basename(mo_path)
            print(f'✅ Created: {mo_name} ({size:,} bytes)')
            print(f'   Translations: {len(translations)}\n')
            
            compiled += 1
        except Exception as e:
            print(f'❌ Error: {str(e)}\n')
    
    print('========================================')
    print(f'Compilation completed!')
    print(f'Total compiled: {compiled} file(s)')
    print('========================================')
    
    return 0 if compiled > 0 else 1

if __name__ == '__main__':
    sys.exit(main())
